KsApi/PKCE.swift

```diff
@@ -0,0 +1,58 @@
+import CryptoKit
+import Foundation
+
+/// PKCE stands for Proof Key for Code Exchange.
+/// The code verifier, and its associated challenge, are used to ensure that an oAuth request is valid.
+/// See this documentation for more details: https://www.oauth.com/oauth2-servers/mobile-and-native-apps/authorization/
+public struct PKCE {
+  /// Creates a random alphanumeric string of the specified length
+  static func createCodeVerifier(ofLength length: Int) throws -> String {
+    let uppercase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    let lowercase = "abcdefghijklmnopqrstuvwxyz"
+    let numbers = "0123456789"
+    let specials = "-._~"
+
+    let validCharacters = uppercase + lowercase + numbers + specials
+
+    var codeVerifier = ""
+
+    for _ in 0..<length {
+      guard let character = validCharacters.randomElement() else {
+        throw ErrorType.UnableToCreateCodeVerifier
+      }
+
+      codeVerifier += String(character)
+    }
+
+    return codeVerifier
+  }
+
+  /// Creates a base-64 encoded SHA256 hash of the given string.
+  static func createCodeChallenge(fromVerifier verifier: String) throws -> String {
+    guard let stringData = verifier.data(using: .utf8) else {
+      throw ErrorType.UnableToCreateCodeChallenge
+    }
+
+    let hash = SHA256.hash(data: stringData)
+    var hashData: Data?
+
+    hash.withUnsafeBytes { pointer in
+      let dataPointer = pointer.bindMemory(to: UInt8.self)
+      hashData = Data(buffer: dataPointer)
+    }
+
+    guard let unwrappedData = hashData else {
+      throw ErrorType.UnableToCreateCodeChallenge
```

Comment:
Somewhat annoyingly - there are a few cases in this code where I need to unwrap a `nil` coming from a system library. In the interest of making this bulletproof, I'm throwing errors if those unwraps fail - but I never expect that to happen, pragmatically. 

If any part of PKCE fails, I think the correct solution is to abort the login process, and just let the user continue in a logged-out state.